File: ios/oplayer/bitshares/GraphenePublicKey.m
//
//  GraphenePublicKey.m
//
//  Created by SYALON on 13-9-3.
//
//

#import "GraphenePublicKey.h"
#import "ChainObjectManager.h"

@interface GraphenePublicKey()
{
    secp256k1_pubkey_compressed _key;
}
@end

@implementation GraphenePublicKey

-(void)dealloc
{
}

- (id)init
{
    self = [super init];
    if (self)
    {
    }
    return self;
}

+ (id)fromWifPublicKey:(NSString*)wif_public_key
{
    if (!wif_public_key || [wif_public_key isEqualToString:@""]) {
        return nil;
    }
    secp256k1_pubkey_compressed key;
    if (__bts_gen_public_key_from_b58address((const unsigned char*)[wif_public_key UTF8String], (const size_t)[wif_public_key length],
                                             [[ChainObjectManager sharedChainObjectManager].grapheneAddressPrefix length], &key)) {
        return [[self alloc] initWithSecp256k1PublicKey:&key];
    }
    return nil;
}

- (id)initWithSecp256k1PublicKey:(const secp256k1_pubkey_compressed*)public_key
{
    self = [super init];
    if (self)
    {
        assert(public_key);
        memcpy(_key.data, public_key->data, sizeof(_key.data));
    }
    return self;
}

- (id)initWithPrivateKey:(GraphenePrivateKey*)private_key
{
    self = [super init];
    if (self)
    {
        if (!__bts_gen_public_key([private_key getKeyData], &_key)) {
            NSAssert(NO, @"Invalid private key.");
        }
    }
    return self;
}

- (secp256k1_pubkey_compressed*)getKeyData
{
    return &_key;
}

- (NSString*)toWifString
{
    unsigned char output[51+10] = {0, };
    size_t output_size = sizeof(output);
    
    NSString* address_prefix = [ChainObjectManager sharedChainObjectManager].grapheneAddressPrefix;
    __bts_public_key_to_address(&_key, output, &output_size, [address_prefix UTF8String], address_prefix.length);
    
    return [[NSString alloc] initWithBytes:output length:output_size encoding:NSUTF8StringEncoding];
}

- (GraphenePublicKey*)child:(const digest_sha256*)child
{
    assert(child);
    
    unsigned char message[sizeof(_key.data) + sizeof(child->data)] = {0, };
    
    memcpy(message, _key.data, sizeof(_key.data));
    memcpy(&message[sizeof(_key.data)], child->data, sizeof(child->data));
    
    digest_sha256 offset = {0, };
    sha256(message, sizeof(message), offset.data);
    
    //  计算 child 公钥
    GraphenePublicKey* child_public_key = [[GraphenePublicKey alloc] initWithSecp256k1PublicKey:[self getKeyData]];
    __bts_pubkey_tweak_add([child_public_key getKeyData], offset.data);
    return child_public_key;
}

/*
 *  (public) 生成变形的 to_public_key，仅用做验证。没发计算出原 to_public_key。
 */
- (GraphenePublicKey*)genToToTo:(NSData*)commitment
{
    assert(commitment);
    assert(commitment.length == 33);
    
    unsigned char buf[sizeof(_key.data) + 33];
    memcpy(buf, _key.data, sizeof(_key.data));
    memcpy(&buf[sizeof(_key.data)], commitment.bytes, 33);
    
    secp256k1_prikey to_digest;
    sha256(buf, sizeof(buf), to_digest.data);
    
    return [[[GraphenePrivateKey alloc] initWithSecp256k1PrivateKey:&to_digest] getPublicKey];
}

@end

File: android/app/src/main/java/bitshares/GraphenePublicKey.kt
package bitshares

import com.btsplusplus.fowallet.NativeInterface
import com.fowallet.walletcore.bts.ChainObjectManager

class GraphenePublicKey {

    companion object {

        fun fromWifPublicKey(wif_public_key: String?): GraphenePublicKey? {
            if (wif_public_key == null || wif_public_key.isEmpty()) {
                return null
            }
            val keydata = NativeInterface.sharedNativeInterface().bts_gen_public_key_from_b58address(wif_public_key.utf8String(),
                    ChainObjectManager.sharedChainObjectManager().grapheneAddressPrefix.utf8String())
            return if (keydata != null) {
                GraphenePublicKey().initWithSecp256k1PublicKey(keydata)
            } else {
                null
            }
        }

    }

    private var _key_data: ByteArray? = null

    fun initWithSecp256k1PublicKey(public_keydata: ByteArray): GraphenePublicKey {
        _key_data = public_keydata
        return this
    }

    fun initWithPrivateKey(private_key: GraphenePrivateKey): GraphenePublicKey {
        return GraphenePublicKey().initWithSecp256k1PublicKey(NativeInterface.sharedNativeInterface().bts_gen_public_key(private_key.getKeyData())!!)
    }

    fun getKeyData(): ByteArray {
        return _key_data!!
    }

    fun toWifString(): String {
        return NativeInterface.sharedNativeInterface().bts_public_key_to_address(_key_data!!,
                ChainObjectManager.sharedChainObjectManager().grapheneAddressPrefix.utf8String())!!.utf8String()
    }

    fun child(child: ByteArray): GraphenePublicKey {
        //  计算 offset
        val offset = sha256(_key_data!! + child)

        //  计算 child 公钥
        return GraphenePublicKey().initWithSecp256k1PublicKey(NativeInterface.sharedNativeInterface().bts_pubkey_tweak_add(_key_data!!, offset)!!)
    }

    /**
     *  (public) 生成变形的 to_public_key，仅用做验证。没发计算出原 to_public_key。
     */
    fun genToToTo(commitment: ByteArray): GraphenePublicKey {
        assert(commitment.size == 33)
        val to_digest = sha256(_key_data!! + commitment)
        return GraphenePrivateKey().initWithSecp256k1PrivateKey(to_digest).getPublicKey()
    }

}